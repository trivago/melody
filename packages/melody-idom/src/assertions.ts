/**
 * Copyright 2015 The Incremental DOM Authors.
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Keeps track whether or not we are in an attributes declaration (after
 * elementOpenStart, but before elementOpenEnd).
 * @type {boolean}
 */
var inAttributes = false;

/**
 * Keeps track whether or not we are in an element that should not have its
 * children cleared.
 * @type {boolean}
 */
var inSkip = false;

/**
 * Makes sure that there is a current patch context.
 * @param {*} context
 */
var assertInPatch = function(functionName, context) {
    if (!context) {
        throw new Error('Cannot call ' + functionName + '() unless in patch');
    }
};

/**
 * Makes sure that a patch closes every node that it opened.
 * @param {?Node} openElement
 * @param {!Node|!DocumentFragment} root
 */
var assertNoUnclosedTags = function(openElement, root) {
    if (openElement === root) {
        return;
    }

    var currentElement = openElement;
    var openTags = [];
    while (currentElement && currentElement !== root) {
        openTags.push(currentElement.nodeName.toLowerCase());
        currentElement = currentElement.parentNode;
    }

    throw new Error(
        'One or more tags were not closed:\n' + openTags.join('\n')
    );
};

/**
 * Makes sure that the caller is not where attributes are expected.
 * @param {string} functionName
 */
var assertNotInAttributes = function(functionName) {
    if (inAttributes) {
        throw new Error(
            functionName +
                '() can not be called between ' +
                'elementOpenStart() and elementOpenEnd().'
        );
    }
};

/**
 * Makes sure that the caller is not inside an element that has declared skip.
 * @param {string} functionName
 */
var assertNotInSkip = function(functionName) {
    if (inSkip) {
        throw new Error(
            functionName +
                '() may not be called inside an element ' +
                'that has called skip().'
        );
    }
};

/**
 * Makes sure that the caller is where attributes are expected.
 * @param {string} functionName
 */
var assertInAttributes = function(functionName) {
    if (!inAttributes) {
        throw new Error(
            functionName +
                '() can only be called after calling ' +
                'elementOpenStart().'
        );
    }
};

/**
 * Makes sure the patch closes virtual attributes call
 */
var assertVirtualAttributesClosed = function() {
    if (inAttributes) {
        throw new Error(
            'elementOpenEnd() must be called after calling ' +
                'elementOpenStart().'
        );
    }
};

/**
 * Makes sure that tags are correctly nested.
 * @param {string} nodeName
 * @param {string} tag
 */
var assertCloseMatchesOpenTag = function(nodeName, tag) {
    if (nodeName !== tag) {
        throw new Error(
            'Received a call to close "' +
                tag +
                '" but "' +
                nodeName +
                '" was open.'
        );
    }
};

/**
 * Makes sure that no children elements have been declared yet in the current
 * element.
 * @param {string} functionName
 * @param {?Node} previousNode
 */
var assertNoChildrenDeclaredYet = function(functionName, previousNode) {
    if (previousNode !== null) {
        throw new Error(
            functionName +
                '() must come before any child ' +
                'declarations inside the current element.'
        );
    }
};

/**
 * Checks that a call to patchOuter actually patched the element.
 * @param {?Node} node The node requested to be patched.
 * @param {?Node} previousNode The previousNode after the patch.
 */
var assertPatchElementNoExtras = function(
    startNode,
    currentNode,
    expectedNextNode,
    expectedPrevNode
) {
    const wasUpdated =
        currentNode.nextSibling === expectedNextNode &&
        currentNode.previousSibling === expectedPrevNode;
    const wasChanged =
        currentNode.nextSibling === startNode.nextSibling &&
        currentNode.previousSibling === expectedPrevNode;
    const wasRemoved = currentNode === startNode;

    if (!wasUpdated && !wasChanged && !wasRemoved) {
        throw new Error(
            'There must be exactly one top level call corresponding ' +
                'to the patched element.'
        );
    }
};

/**
 * Updates the state of being in an attribute declaration.
 * @param {boolean} value
 * @return {boolean} the previous value.
 */
var setInAttributes = function(value) {
    var previous = inAttributes;
    inAttributes = value;
    return previous;
};

/**
 * Updates the state of being in a skip element.
 * @param {boolean} value
 * @return {boolean} the previous value.
 */
var setInSkip = function(value) {
    var previous = inSkip;
    inSkip = value;
    return previous;
};

/** */
export {
    assertInPatch,
    assertNoUnclosedTags,
    assertNotInAttributes,
    assertInAttributes,
    assertCloseMatchesOpenTag,
    assertVirtualAttributesClosed,
    assertNoChildrenDeclaredYet,
    assertNotInSkip,
    assertPatchElementNoExtras,
    setInAttributes,
    setInSkip,
};
