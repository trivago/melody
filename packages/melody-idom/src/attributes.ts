/**
 * Copyright 2015 The Incremental DOM Authors.
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getData } from './node_data';

const getNamespace = function(name) {
    if (name.lastIndexOf('xml:', 0) === 0) {
        return 'http://www.w3.org/XML/1998/namespace';
    }

    if (name.lastIndexOf('xlink:', 0) === 0) {
        return 'http://www.w3.org/1999/xlink';
    }
};

/**
 * Applies an attribute or property to a given Element. If the value is null
 * or undefined, it is removed from the Element. Otherwise, the value is set
 * as an attribute.
 * @param {!Element} el
 * @param {string} name The attribute's name.
 * @param {?(boolean|number|string)=} value The attribute's value.
 */
const applyAttr = function(el, name, value) {
    if (value == null) {
        el.removeAttribute(name);
    } else {
        const attrNS = getNamespace(name);
        if (attrNS) {
            el.setAttributeNS(attrNS, name, value);
        } else {
            el.setAttribute(name, value);
        }
    }
};

/**
 * Updates a single attribute on an Element.
 * @param {!Element} el
 * @param {string} name The attribute's name.
 * @param {*} value The attribute's value. If the value is an object or
 *     function it is set on the Element, otherwise, it is set as an HTML
 *     attribute.
 */
function applyAttributeTyped(el, name, value) {
    var type = typeof value;

    if (type === 'object' || type === 'function') {
        setProperty(el, name, value);
    } else {
        applyAttr(el, name /** @type {?(boolean|number|string)} */, value);
    }
}

function setProperty(el, name, value) {
    try {
        el[name] = value;
    } catch (e) {}
}

function eventProxy(e) {
    return this._listeners[e.type](e);
}

/**
 * Calls the appropriate attribute mutator for this attribute.
 * @param {!Element} el
 * @param {string} name The attribute's name.
 * @param {*} value The attribute's value.
 */
const updateAttribute = function(el, name, value) {
    var data = getData(el);
    var attrs = data.attrs;

    if (attrs[name] === value) {
        return;
    }

    if (name === 'style') {
        const old = attrs.style;
        if (!value || typeof value === 'string') {
            el.style.cssText = value || '';
        } else {
            if (typeof old === 'string') {
                el.style.cssText = '';
            } else {
                for (let i in old) {
                    if (!(i in value)) {
                        el.style[i] = '';
                    }
                }
            }
            for (let i in value) {
                if (i.indexOf('-') >= 0) {
                    el.style.setProperty(i, value[i]);
                } else {
                    el.style[i] = value[i];
                }
            }
        }
    } else if (name === 'ref') {
        const old = attrs.ref;
        if (old && old.disposer) {
            if (old.creator === value) {
                return;
            }
            old.disposer.unsubscribe();
        }
        if (!value) {
            attrs.ref = null;
            return;
        }
        attrs.ref = {
            creator: value,
            disposer: value(el),
        };
        if (process.env.NODE_ENV !== 'production') {
            if (
                !attrs.ref.disposer ||
                typeof attrs.ref.disposer.unsubscribe !== 'function'
            ) {
                throw new Error(
                    `A ref handler is supposed to return a Subscription object which must have a "unsubscribe" method.`
                );
            }
        }
        return;
    } else if (name[0] === 'o' && name[1] === 'n') {
        if (typeof value === 'string') {
            applyAttributeTyped(el, name, value);
        } else {
            let eventName = name.replace(/Capture$/, '');
            let useCapture = name !== eventName;
            eventName = eventName.toLowerCase().substring(2);
            if (value) {
                if (!attrs[name]) {
                    el.addEventListener(eventName, eventProxy, useCapture);
                }
            } else if (typeof attrs[name] === 'string') {
                el.removeAttribute(name);
            } else {
                el.removeEventListener(eventName, eventProxy, useCapture);
            }
            (el._listeners || (el._listeners = {}))[eventName] = value;
        }
    } else if (
        name !== 'list' &&
        name !== 'type' &&
        name !== 'draggable' &&
        !(el.ownerSVGElement || el.localName === 'svg') &&
        name in el
    ) {
        setProperty(el, name, value == null ? '' : value);
        if (value == null || value === false) {
            el.removeAttribute(name);
        }
    } else {
        applyAttributeTyped(el, name, value);
    }

    attrs[name] = value;
};

/** */
export { updateAttribute };
