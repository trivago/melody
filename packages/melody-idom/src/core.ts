/**
 * Copyright 2015 The Incremental DOM Authors.
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { createElement, createText, createRaw } from './nodes';
import { getData } from './node_data';
import {
    assertInPatch,
    assertNoUnclosedTags,
    assertNotInAttributes,
    assertVirtualAttributesClosed,
    assertNoChildrenDeclaredYet,
    assertPatchElementNoExtras,
    setInAttributes,
    setInSkip,
} from './assertions';
import { getFocusedPath, moveBefore } from './dom_util';
import { unmountComponent } from './util';
import { enqueueComponent } from './renderQueue';
import { link, reset } from './hierarchy';

type Class<T> = { new (): T };

export interface CallableComponent {
    apply(props: any): void;
    el: Element;
}

export interface RenderableComponent {
    el: Element;
    refs: any;
    render(): void;
    notify(): void;
    type: String;
}

/** @type {?Node} */
var currentNode;

/** @type {?Node} */
var currentParent;

/** @type {?Document} */
var doc;

var componentKey = null;
var currentComponent = null;

var deletedNodes: Array<Node> = null;

const markFocused = function(focusPath: Array<Node>, focused: boolean): void {
    for (let i = 0; i < focusPath.length; i += 1) {
        getData(focusPath[i]).focused = focused;
    }
};

const patchFactory = function(run) {
    return function(node, fn, data) {
        if (process.env.NODE_ENV !== 'production') {
            if (!node) {
                throw new Error('Patch invoked without an element.');
            }
        }
        const prevDeletedNodes = deletedNodes;
        const prevDoc = doc;
        const prevCurrentNode = currentNode;
        const prevCurrentParent = currentParent;
        const prevCurrentComponent = currentComponent;

        let previousInAttribute = false;
        let previousInSkip = false;

        deletedNodes = [];
        doc = node.ownerDocument;
        currentParent = node.parentNode;

        if (process.env.NODE_ENV !== 'production') {
            previousInAttribute = setInAttributes(false);
            previousInSkip = setInSkip(false);
        }

        const focusPath = getFocusedPath(node, currentParent);
        markFocused(focusPath, true);
        let retVal;
        if (process.env.NODE_ENV !== 'production') {
            try {
                retVal = run(node, fn, data);
            } catch (e) {
                // reset context
                deletedNodes = prevDeletedNodes;
                doc = prevDoc;
                currentNode = prevCurrentNode;
                currentParent = prevCurrentParent;
                currentComponent = prevCurrentComponent;
                // rethrow the error
                throw e;
            }
        } else {
            retVal = run(node, fn, data);
        }

        markFocused(focusPath, false);

        if (process.env.NODE_ENV !== 'production') {
            assertVirtualAttributesClosed();
            setInAttributes(previousInAttribute);
            setInSkip(previousInSkip);
        }

        var i, len;
        for (i = 0, len = deletedNodes.length; i < len; i++) {
            nodeDeleted(deletedNodes[i]);
        }

        // reset context
        deletedNodes = prevDeletedNodes;
        doc = prevDoc;
        currentNode = prevCurrentNode;
        currentParent = prevCurrentParent;
        currentComponent = prevCurrentComponent;

        return retVal;
    };
};

function nodeDeleted(node) {
    const data = getData(node);
    if (data.attrs.ref && data.attrs.ref.disposer) {
        data.attrs.ref.disposer.unsubscribe();
        data.attrs.ref = null;
    }
    if (data.componentInstance) {
        unmountComponent(data.componentInstance);
    }
    // not an ideal solution but we can eventually move it
    // towards a scheduler (perhaps `requestIdleCallback` if we notice
    // that there are actual issues with this)
    // Chose a recursive solution here to avoid unnecessary memory usage
    let child = node.firstChild;
    while (child) {
        nodeDeleted(child);
        child = child.nextSibling;
    }
}

const patchInner = patchFactory(function(node, fn, data) {
    currentNode = node;
    enterNode();
    fn(data);
    exitNode();

    if (process.env.NODE_ENV !== 'production') {
        assertNoUnclosedTags(currentNode, node);
    }

    return node;
});

const patchOuter = patchFactory(function(node, fn, data) {
    const startNode = { nextSibling: node };
    let expectedNextNode = null;
    let expectedPrevNode = null;

    if (process.env.NODE_ENV !== 'production') {
        expectedNextNode = node.nextSibling;
        expectedPrevNode = node.previousSibling;
    }

    currentNode = startNode;
    fn(data);

    if (process.env.NODE_ENV !== 'production') {
        assertPatchElementNoExtras(
            startNode,
            currentNode,
            expectedNextNode,
            expectedPrevNode
        );
    }

    if (node !== currentNode && node.parentNode) {
        removeChild(currentParent, node, getData(currentParent).keyMap);
    }

    return startNode === currentNode ? null : currentNode;
});

/**
 * Checks whether or not the current node matches the specified nodeName and
 * key.
 *
 * @param {?string} nodeName The nodeName for this node.
 * @param {?string=} key An optional key that identifies a node.
 * @return {boolean} True if the node matches, false otherwise.
 */
var matches = function(
    matchNode: Node,
    nodeName: string,
    key?: string
): boolean {
    var data = getData(matchNode);

    // Key check is done using double equals as we want to treat a null key the
    // same as undefined. This should be okay as the only values allowed are
    // strings, null and undefined so the == semantics are not too weird.
    // templates rendered on the server side may not have keys at all while melody templates
    // always will have them so we reconcile the dom in those cases.
    if (nodeName === data.nodeName) {
        if (key == data.key) {
            return true;
        }
        // exisiting DOM element does not have a key
        // which means we can hook onto it freely
        if (!data.key) {
            data.key = key;
            data.staticsApplied = false;
            // but we'll need to update the parent element
            const parentKeys = currentParent && getData(currentParent).keyMap;
            if (parentKeys) {
                parentKeys[key] = matchNode;
            }
            return true;
        }
    }
    return false;
};

/**
 * Aligns the virtual Element definition with the actual DOM, moving the
 * corresponding DOM node to the correct location or creating it if necessary.
 * @param {string} nodeName For an Element, this should be a valid tag string.
 *     For a Text, this should be #text.
 * @param {?string=} key The key used to identify this element.
 * @param {?Array<*>=} statics For an Element, this should be an array of
 *     name-value pairs.
 */
var alignWithDOM = function(nodeName: string, key?: string): void {
    if (currentNode && matches(currentNode, nodeName, key)) {
        return;
    }

    const parentData = getData(currentParent);
    const currentNodeData = currentNode && getData(currentNode);
    const keyMap = parentData.keyMap;
    let fromKeyMap = false;
    let node;
    let componentInstance = null;

    // Check to see if the node has moved within the parent.
    if (key) {
        const keyNode = keyMap[key];
        if (keyNode) {
            if (matches(keyNode, nodeName, key)) {
                fromKeyMap = true;
                node = keyNode;
            } else if (keyNode === currentNode) {
                const keyNodeData = getData(keyNode);
                // if (keyNodeData.componentInstance === currentComponent) {
                if (keyNodeData.componentInstance) {
                    componentInstance = keyNodeData.componentInstance;
                    keyNodeData.componentInstance = null;
                } else {
                    deletedNodes.push(keyNode);
                }
            } else {
                removeChild(currentParent, keyNode, keyMap);
            }
        } else if (
            currentNode &&
            currentNode.nodeType === 3 &&
            currentNode.data.trim() === ''
        ) {
            // special handling here to ignore empty text nodes if the one after it is what we're actually looking for
            // this reduces a lot of special handling for server side rendered content.
            if (
                currentNode.nextSibling &&
                matches(currentNode.nextSibling, nodeName, key)
            ) {
                node = currentNode.nextSibling;
            }
        }
    }

    // Create the node if it doesn't exist.
    if (!node) {
        if (nodeName === '#text') {
            node = createText(doc);
        } else {
            node = createElement(doc, currentParent, nodeName, key);
        }

        if (key) {
            keyMap[key] = node;
        }
    }

    if (componentInstance) {
        getData(node).componentInstance = componentInstance;
        componentInstance.el = node;
    }

    // Re-order the node into the right position, preserving focus if either
    // node or currentNode are focused by making sure that they are not detached
    // from the DOM.
    if (getData(node).focused) {
        // move everything else before the node.
        moveBefore(currentParent, node, currentNode);
    } else if (
        !(fromKeyMap && !node.parentNode) &&
        currentNodeData &&
        currentNodeData.key &&
        !currentNodeData.focused
    ) {
        // Remove the currentNode, which can always be added back since we hold a
        // reference through the keyMap. This prevents a large number of moves when
        // a keyed item is removed or moved backwards in the DOM.
        currentParent.replaceChild(node, currentNode);
        parentData.keyMapValid = false;
    } else if (
        currentNode &&
        currentNode.nextSibling === node &&
        currentNode.nodeType === 3 &&
        currentNode.data.trim() === ''
    ) {
        // if the empty text node handling above was successful, we simply remove the skipped text node
        currentParent.removeChild(currentNode);
    } else {
        currentParent.insertBefore(node, currentNode);
    }

    currentNode = node;
};

const removeChild = function(node: Node, child: Node, keyMap): void {
    node.removeChild(child);
    deletedNodes.push(child);

    const key = getData(child).key;
    if (key) {
        delete keyMap[key];
    }
};

/**
 * Clears out any unvisited Nodes, as the corresponding virtual element
 * functions were never called for them.
 */
var clearUnvisitedDOM = function(): void {
    var node = currentParent;
    var data = getData(node);
    var keyMap = data.keyMap;
    var keyMapValid = data.keyMapValid;
    var child = node.lastChild;
    var key;

    if (child === currentNode && keyMapValid) {
        return;
    }

    while (child && child !== currentNode) {
        removeChild(node, child, keyMap);
        child = node.lastChild;
    }

    // Clean the keyMap, removing any unusued keys.
    if (!keyMapValid) {
        for (key in keyMap) {
            child = keyMap[key];
            if (child.parentNode !== node) {
                deletedNodes.push(child);
                delete keyMap[key];
            }
        }

        data.keyMapValid = true;
    }
};

/**
 * Changes to the first child of the current node.
 */
var enterNode = function(): void {
    currentParent = currentNode;
    currentNode = null;
};

/**
 * Changes to the next sibling of the current node.
 */
var nextNode = function(): void {
    currentNode = getNextNode();
};

var getNextNode = function(): Node | null {
    if (currentNode) {
        return currentNode.nextSibling;
    } else {
        return currentParent.firstChild;
    }
};

/**
 * Changes to the parent of the current node, removing any unvisited children.
 */
var exitNode = function(): void {
    clearUnvisitedDOM();

    currentNode = currentParent;
    currentParent = currentParent.parentNode;
};

var updateComponent = function(comp: RenderableComponent): void {
    const data = getData(comp.el);
    const parentComponent = currentComponent;
    componentKey = data.key;

    reset(comp);

    currentComponent = comp;
    comp.render();

    currentComponent = parentComponent;
};

var scheduleComponent = function(
    Component: Class<CallableComponent> | CallableComponent,
    key: string,
    props: any,
    el?: Node
): any {
    var comp;
    if (el) {
        // we've already seen this component
        var data = getData(el);
        comp = data.componentInstance;
        if (!comp) {
            // but apparently we didn't have a component instance so far
            // most likely we're mounting a server side rendered DOM
            comp =
                typeof Component === 'function' ? new Component() : Component;
            comp.el = el;
            data.componentInstance = comp;
        }
        // Q: Do we even want to support this in the future?
        // if (typeof Component === 'function' && !(comp instanceof Component)) {
        //   unmountComponent(comp);
        //   comp = null;
        // }
        elementOpen(data.nodeName, key);
        skip();
        elementClose();
    } else {
        // unknown component
        if (typeof Component === 'function') {
            comp = new Component();
        } else {
            comp = Component;
        }

        elementOpen('m-placeholder', key);
        skip();
        comp.el = elementClose();
        getData(comp.el).componentInstance = comp;
    }

    if (currentComponent) {
        link(currentComponent, comp);
    }
    return comp.apply(props);
};

var component = function(
    Component: Class<CallableComponent> | CallableComponent,
    key: string,
    props: any
): any {
    var el = getData(currentParent).keyMap[key];
    return scheduleComponent(Component, key, props, el);
};

var getCurrentComponent = function(): RenderableComponent {
    return currentComponent;
};

var mount = function(
    element: Node,
    Component: Class<CallableComponent> | CallableComponent,
    props: any
): any {
    var data = getData(element);
    var key = data && data.key;
    var comp = data.componentInstance;
    var isComponentInstance = typeof Component !== 'function';
    // if the existing component is not an instance of the specified component type
    // then we just unmount the existing one and proceed as if none ever existed
    if (
        comp &&
        !isComponentInstance &&
        !(comp instanceof (Component as Class<CallableComponent>))
    ) {
        unmountComponent(comp);
    }
    return scheduleComponent(Component, key, props, element);
};

/**
 * Makes sure that the current node is an Element with a matching tagName and
 * key.
 *
 * @param {string} tag The element's tag.
 * @param {?string=} key The key used to identify this element. This can be an
 *     empty string, but performance may be better if a unique value is used
 *     when iterating over an array of items.
 * @return {!Element} The corresponding Element.
 */
var elementOpen = function(tag: string, key?: string) {
    nextNode();
    alignWithDOM(tag, componentKey || key);
    componentKey = null;
    enterNode();
    return currentParent;
};

/**
 * Closes the currently open Element, removing any unvisited children if
 * necessary.
 *
 * @return {!Element} The corresponding Element.
 */
var elementClose = function(): void {
    if (process.env.NODE_ENV !== 'production') {
        setInSkip(false);
    }

    exitNode();
    return currentNode;
};

/**
 * Makes sure the current node is a Text node and creates a Text node if it is
 * not.
 *
 * @return {!Text} The corresponding Text Node.
 */
var text = function(): Text {
    nextNode();
    alignWithDOM('#text', null);
    return currentNode;
};

/**
 * Gets the current Element being patched.
 * @return {!Element}
 */
var currentElement = function(): Node {
    if (process.env.NODE_ENV !== 'production') {
        assertInPatch('currentElement', deletedNodes);
        assertNotInAttributes('currentElement');
    }
    return currentParent;
};

/**
 * Skips the children in a subtree, allowing an Element to be closed without
 * clearing out the children.
 */
var skip = function(): void {
    if (process.env.NODE_ENV !== 'production') {
        assertNoChildrenDeclaredYet('skip', currentNode);
        setInSkip(true);
    }
    currentNode = currentParent.lastChild;
};

var inPatch = function(): boolean {
    return !!deletedNodes;
};

const skipNode = nextNode;

var insertRawHtml = function(html: string): void {
    var children = createRaw(doc, html);
    var node = doc.createDocumentFragment(),
        lastChild = children[children.length - 1];
    while (children.length) {
        node.appendChild(children[0]);
    }
    currentParent.insertBefore(node, currentNode);
    currentNode = lastChild;
};

var raw = function(html: string): void {
    nextNode();
    if (currentNode && matches(currentNode, '#raw', null)) {
        // patch node
        var data = getData(currentNode),
            remainingSiblingCount = data.childLength - 1;
        if (data.text !== html) {
            // if the text is not the same as before, we'll have some work to do
            insertRawHtml(html);
            // remove the remaining siblings of the old child
            if (data.childLength > 1) {
                while (remainingSiblingCount--) {
                    currentParent.removeChild(currentNode.nextSibling);
                }
            }
        } else if (remainingSiblingCount) {
            // still the same text so just jump over the remaining siblings
            while (remainingSiblingCount--) {
                currentNode = currentNode.nextSibling;
            }
        }
    } else {
        // insert raw html
        insertRawHtml(html);
    }

    return currentNode;
};

/** */
export {
    elementOpen,
    elementClose,
    text,
    patchInner,
    patchOuter,
    currentElement,
    skip,
    skipNode,
    inPatch,
    raw,
    mount,
    component,
    getCurrentComponent as currentComponent,
    updateComponent,
    enqueueComponent,
};
