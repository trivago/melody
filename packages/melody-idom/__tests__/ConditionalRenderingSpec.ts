/**
 * Copyright 2015 The Incremental DOM Authors.
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { patch, elementOpen, elementClose, elementVoid } from '../src';

describe('conditional rendering', () => {
    let container;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });

    afterEach(() => {
        document.body.removeChild(container);
    });

    describe('nodes', () => {
        function render(condition) {
            elementOpen('div', 'outer', ['id', 'outer']);
            elementVoid('div', 'one', ['id', 'one']);

            if (condition) {
                elementVoid('div', 'conditional-one', [
                    'id',
                    'conditional-one',
                ]);
                elementVoid('div', 'conditional-two', [
                    'id',
                    'conditional-two',
                ]);
            }

            elementVoid('span', 'two', ['id', 'two']);
            elementClose('div');
        }

        it('should un-render when the condition becomes false', () => {
            patch(container, () => render(true));
            patch(container, () => render(false));
            const outer = container.childNodes[0];

            expect(outer.childNodes).toHaveLength(2);
            expect(outer.childNodes[0].id).toEqual('one');
            expect(outer.childNodes[0].tagName).toEqual('DIV');
            expect(outer.childNodes[1].id).toEqual('two');
            expect(outer.childNodes[1].tagName).toEqual('SPAN');
        });

        it('should render when the condition becomes true', () => {
            patch(container, () => render(false));
            patch(container, () => render(true));
            const outer = container.childNodes[0];

            expect(outer.childNodes).toHaveLength(4);
            expect(outer.childNodes[0].id).toEqual('one');
            expect(outer.childNodes[0].tagName).toEqual('DIV');
            expect(outer.childNodes[1].id).toEqual('conditional-one');
            expect(outer.childNodes[1].tagName).toEqual('DIV');
            expect(outer.childNodes[2].id).toEqual('conditional-two');
            expect(outer.childNodes[2].tagName).toEqual('DIV');
            expect(outer.childNodes[3].id).toEqual('two');
            expect(outer.childNodes[3].tagName).toEqual('SPAN');
        });
    });

    describe('with only conditional childNodes', () => {
        function render(condition) {
            elementOpen('div', 'outer', ['id', 'outer']);

            if (condition) {
                elementVoid('div', 'conditional-one', [
                    'id',
                    'conditional-one',
                ]);
                elementVoid('div', 'conditional-two', [
                    'id',
                    'conditional-two',
                ]);
            }

            elementClose('div');
        }

        it('should not leave any remaning nodes', () => {
            patch(container, () => render(true));
            patch(container, () => render(false));
            const outer = container.childNodes[0];

            expect(outer.childNodes).toHaveLength(0);
        });
    });

    describe('with static attributes', () => {
        const _statics = ["class", "foo"];
        function render(_context) {
            elementOpen("div", null, null);

             if (_context.condition) {
                elementVoid("div", "7*U2;JR", _statics);
            } else {
                elementVoid("div", null, null, "class", "foo " + (_context.bar ? "bar" : ""));
            }

             elementClose("div");
        }

         it('should apply static attributes when recycling an element', () => {
            patch(container, () => render({ condition: false, bar: true }));
            expect(container.innerHTML).toMatchSnapshot();
            patch(container, () => render({ condition: true, bar: true }));
            expect(container.innerHTML).toMatchSnapshot();
        });

         it('should remove static attributes when recycling an element', () => {
            patch(container, () => render({ condition: true, bar: true }));
            expect(container.innerHTML).toMatchSnapshot();
            patch(container, () => render({ condition: false, bar: true }));
            expect(container.innerHTML).toMatchSnapshot();
        });
    });

    describe('nodes', () => {
        function render(condition) {
            elementOpen('div', null, null, 'id', 'outer');
            elementVoid('div', null, null, 'id', 'one');

            if (condition) {
                elementOpen(
                    'span',
                    'conditional',
                    null,
                    'id',
                    'conditional-one',
                    'data-foo',
                    'foo'
                );
                elementVoid('span');
                elementClose('span');
            }

            elementVoid('span', 'last', null, 'id', 'two');
            elementClose('div');
        }

        it('should strip children when a conflicting node is re-used', () => {
            patch(container, () => render(true));
            patch(container, () => render(false));
            const outer = container.childNodes[0];

            expect(outer.childNodes).toHaveLength(2);
            expect(outer.childNodes[0].id).toEqual('one');
            expect(outer.childNodes[0].tagName).toEqual('DIV');
            expect(outer.childNodes[1].id).toEqual('two');
            expect(outer.childNodes[1].tagName).toEqual('SPAN');
            expect(outer.childNodes[1].children.length).toEqual(0);
        });

        it('should strip attributes when a conflicting node is re-used', () => {
            patch(container, () => render(true));
            patch(container, () => render(false));
            const outer = container.childNodes[0];

            expect(outer.childNodes[1].getAttribute('data-foo')).toBeNull();
        });
    });
});
