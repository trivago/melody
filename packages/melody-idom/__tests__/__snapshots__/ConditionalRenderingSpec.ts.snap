// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditional rendering with static attributes should apply static attributes when recycling an element 1`] = `"<div><div class=\\"foo bar\\"></div></div>"`;

exports[`conditional rendering with static attributes should apply static attributes when recycling an element 2`] = `"<div><div class=\\"foo\\"></div></div>"`;

exports[`conditional rendering with static attributes should remove static attributes when recycling an element 1`] = `"<div><div class=\\"foo\\"></div></div>"`;

exports[`conditional rendering with static attributes should remove static attributes when recycling an element 2`] = `"<div><div class=\\"foo bar\\"></div></div>"`;

